### ISSUE TITLE
Add Script for replacing MegaMan with Hatsune Miku

### ISSUE THREAD

**Issue by kion-dgl** [DashGL/Tool-MML2SaveStateViewer#6]
While I've been doing a lot of dirty proof of concepts for getting Hatsune Miku into the game, the next step is to clean up the code that I have in order to make a script that does replacement. 

The script is going to need three inputs

1. The updated assets in a common format (.png, .obj)
2. The gamefiles to be updated (.BIN, .DAT)
3. The game ROM 

While it's not in the scope of this issue, it might be helpful to think about how this code would work if it supported both PSX and PC.

- First we need a folder with assets with specific names. We read the assets from that folder, and then encode them into the game's format. 
- Then we replace the assets in their respective archives. And then we can write these to disk
  - In the case of PC this is where we would be done
  - In the case of PSX, this is where we would need to search for the original file in the ROM and replace it

The main thing that my brain is trying to sort out right now is the data table that we're going to need to set up to match the updated assets, with the files they'll be replacing and their locations in the files. 

The easiest way to do this is probably going to be to have one long JSON file, where were store all of these relations. These relations can then be filtered into class instances that know how to handle them.

**Comment by kion-dgl** [comment]
First pass on this. Managed to get the Miku texture encoded, compressed and applied to MegaMan directly. 

![image](https://github.com/DashGL/Dash2-Toolkit/assets/25621780/f0fac15e-2114-44f8-b4d8-3469632cde7d)

What I was surprised about was only the body texture was encoded and loaded into the game. The order of `PL00T.BIN` goes body texture, alt body palette and face texture. Here we a somewhat expected result in that the body texture loaded in without crashing, but not the textures that would normally follow it.

Which means processing on the file likely ended for some reason.

Looking into updated file, it wasn't too hard to figure out why. Our compression function has slightly higher compression than the one used in the game. Which means our compressed body image ended earlier than expected in the `.BIN` archive. 

The game then expects the next file in the archive start at the next multiple of 0x400. And since we're just under that limit, the archive loader function terminates. This presents some interesting but complicated possibilities for modding, in that it looks like the archive formats could be used to deliver a payload into memory as long as there is room in the file.

Something that we are likely to never take advantage of, but something to keep in mind just in case. 

For now, what this means is that our best approach is likely to rebuild the file in whole. We can copy the header from the source file, and then stack them on top of each other in the resulting archive. And that should allow us to proceed with injecting the face texture. 

- Copy header from image 0
- Copy palette
- Copy header from image 1
- Update header for image 0 and image 1
- Stack the file in order

One last thing that I'm musing about in this thread is the approach for managing each file. It's possible that a one size fits all approach might not work and I might need a per-file approach so that each file knows what it needs to do. 

It also makes sense to add in the src and out for the PC files early on to be able to account for both, and export both from the script.

**Comment by kion-dgl** [comment]
In terms of progress, we've managed to write code for encoding the texture for the body and face, encoding it, and then replacing it in the PSX and PC file. In the next video that I create for Megaman Legends, I think I want the focus to be on compression. 

For the mod, the next aspect to focus on is the code for encoding the mesh. I should be able to do what I did previously and start with the body and encode one section at a time and then replace the file at the end. 

One detail that I'm caught up on is the pointers to lighting. I don't think that's something that I'll be using anytime soon. So it seems like a good idea to make a pocket of `0x80` somewhere and keep the pointers constant for all of the headers in the file. 

There are two places that make sense to do this. One is in the section in the feet. And the other is to remove the shoulder from the buster section and use that. I think that rather than trying to get this perfect from the start we can use the buster section as a placeholder to get started and comeback and visit it later. 

For now we can focus on the body and work through encoding one section of the body at a time. Maybe we can get fancy and come up with approaches for what the most efficient way to pack the body is. In general, the only fixed variable is the location of the headers in each respective section. The data in-between all of these can be packed in any order.
